src/mapreduce.jl

```diff
@@ -86,28 +106,21 @@ end
 ## mapreduce ##
 ###############
 
-@inline function mapreduce(f, op, a::StaticArray, b::StaticArray...; dims=:,kw...)
-    _mapreduce(f, op, dims, kw.data, same_size(a, b...), a, b...)
+@inline function mapreduce(f, op, a::StaticArray, b::StaticArray...; dims=:, init = _InitialValue())
+    _mapreduce(f, op, dims, init, same_size(a, b...), a, b...)
 end
 
-@generated function _mapreduce(f, op, dims::Colon, nt::NamedTuple{()},
-                               ::Size{S}, a::StaticArray...) where {S}
+@inline _mapreduce(args::Vararg{Any,N}) where N = _mapfoldl(args...)
+
+@generated function _mapfoldl(f, op, dims::Colon, init, ::Size{S}, a::StaticArray...) where {S}
     tmp = [:(a[$j][1]) for j ∈ 1:length(a)]
     expr = :(f($(tmp...)))
-    for i ∈ 2:prod(S)
-        tmp = [:(a[$j][$i]) for j ∈ 1:length(a)]
-        expr = :(op($expr, f($(tmp...))))
-    end
-    return quote
-        @_inline_meta
-        @inbounds return $expr
+    if init === _InitialValue
+        expr = :(Base.reduce_first(op, $expr))
```

Comment:
`Base.reduce_first` wasn't called before so `sum(SA[true])` returned `true`; now it returns `1 :: Int`.  This is consistent with `sum([true])`.